File: Source/PokeSharpCore/Private/Interop/PokeSharpTextExporter.cpp
// Fill out your copyright notice in the Description page of Project Settings.

#include "Interop/PokeSharpTextExporter.h"

void UPokeSharpTextExporter::FromSourceString(const UTF16CHAR *Buffer, const int32 Length, FText &OutText)
{
    OutText = FText::FromStringView(TStringView(Buffer, Length));
}

void UPokeSharpTextExporter::FromLocalized(const UTF16CHAR *NamespaceBuffer, const int32 NamespaceLength,
                                           const UTF16CHAR *KeyBuffer, const int32 KeyLength,
                                           const UTF16CHAR *StrBuffer, const int32 StrLength, FText &OutText)
{
    OutText =
        FText::AsLocalizable_Advanced(FTextKey(TStringView(NamespaceBuffer, NamespaceLength)),
                                      FTextKey(TStringView(KeyBuffer, KeyLength)), TStringView(StrBuffer, StrLength));
}

const FString *UPokeSharpTextExporter::GetSourceString(const FText &Text)
{
    return FTextInspector::GetSourceString(Text);
}

const FString *UPokeSharpTextExporter::GetDisplayString(const FText &Text)
{
    return &Text.ToString();
}

bool UPokeSharpTextExporter::GetNamespace(const FText &Text, FString &OutNamespace)
{
    if (auto Result = FTextInspector::GetNamespace(Text); Result.IsSet())
    {
        OutNamespace = MoveTemp(Result.GetValue());
        return true;
    }

    OutNamespace.Empty();
    return false;
}

bool UPokeSharpTextExporter::GetKey(const FText &Text, FString &OutKey)
{
    if (auto Result = FTextInspector::GetKey(Text); Result.IsSet())
    {
        OutKey = MoveTemp(Result.GetValue());
        return true;
    }

    OutKey.Empty();
    return false;
}

bool UPokeSharpTextExporter::IsCultureInvariant(const FText &Text)
{
    return Text.IsCultureInvariant();
}

bool UPokeSharpTextExporter::IsTransient(const FText &Text)
{
    return Text.IsTransient();
}

bool UPokeSharpTextExporter::ShouldGatherForLocalization(const FText &Text)
{
    return Text.ShouldGatherForLocalization();
}

bool UPokeSharpTextExporter::IsEmpty(const FText &Text)
{
    return Text.IsEmpty();
}

bool UPokeSharpTextExporter::IsWhitespace(const FText &Text)
{
    return Text.IsEmptyOrWhitespace();
}

void UPokeSharpTextExporter::Destroy(FText &Text)
{
    std::destroy_at(&Text);
}

void UPokeSharpTextExporter::AsDisplaySpan(const FText &Text, const TCHAR *&OutBuffer, int32 &OutLength)
{
    const auto &AsString = Text.ToString();
    OutBuffer = AsString.GetCharArray().GetData();
    OutLength = AsString.Len();
}

File: Script/PokeSharp.Unreal.Core/Interop/PokeSharpTextExporter.cs
using UnrealSharp.Binds;
using UnrealSharp.Core;

namespace PokeSharp.Unreal.Core.Interop;

[NativeCallbacks]
public static unsafe partial class PokeSharpTextExporter
{
    private static readonly delegate* unmanaged<IntPtr, int, out FTextData, void> FromSourceString;
    private static readonly delegate* unmanaged<
        IntPtr,
        int,
        IntPtr,
        int,
        IntPtr,
        int,
        out FTextData,
        void> FromLocalized;
    private static readonly delegate* unmanaged<ref FTextData, IntPtr> GetSourceString;
    private static readonly delegate* unmanaged<ref FTextData, IntPtr> GetDisplayString;
    private static readonly delegate* unmanaged<ref FTextData, ref UnmanagedArray, NativeBool> GetNamespace;
    private static readonly delegate* unmanaged<ref FTextData, ref UnmanagedArray, NativeBool> GetKey;
    private static readonly delegate* unmanaged<ref FTextData, NativeBool> IsCultureInvariant;
    private static readonly delegate* unmanaged<ref FTextData, NativeBool> IsTransient;
    private static readonly delegate* unmanaged<ref FTextData, NativeBool> ShouldGatherForLocalization;
    private static readonly delegate* unmanaged<ref FTextData, NativeBool> IsEmpty;
    private static readonly delegate* unmanaged<ref FTextData, NativeBool> IsWhitespace;
    private static readonly delegate* unmanaged<ref FTextData, void> Destroy;
    private static readonly delegate* unmanaged<ref FTextData, out IntPtr, out int, void> AsDisplaySpan;
}
